ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,412 @@
+					Ok(logs) => logs,
+					Err(err) => {
+						error!(target: "privatetx", "Cannot deserialize logs from file: {}", err);
+						bail!("Cannot deserialize logs from file: {:?}", err);
+					}
+				};
+				Ok(transaction_logs)
+			}
+			None => {
+				Ok(Vec::new())
+			}
+		}
+	}
+
+	fn flush_logs(&self, logs: &HashMap<H256, TransactionLog>) -> Result<(), Error> {
+		if logs.is_empty() {
+			// Do not create empty file
+			return Ok(());
+		}
+		let log_file = self.open_file(true)?;
+		if let Some(log_file) = log_file {
+			let mut json = serde_json::Serializer::new(log_file);
+			let mut json_array = json.serialize_seq(Some(logs.len()))?;
+			for v in logs.values() {
+				json_array.serialize_element(v)?;
+			}
+			json_array.end()?;
+		}
+		Ok(())
+	}
+}
+
+/// Timestamp source for logs
+pub trait TimestampSource: Send + Sync + 'static {
+	/// Returns current timestamp in seconds
+	fn current_timestamp(&self) -> u64;
+}
+
+/// Timesource on the base of system time
+pub struct SystemTimestamp {}
+
+impl TimestampSource for SystemTimestamp {
+	fn current_timestamp(&self) -> u64 {
+		SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs()
+	}
+}
+
+/// Private transactions logging
+pub struct Logging {
+	logs: RwLock<HashMap<H256, TransactionLog>>,
+	logs_serializer: Arc<LogsSerializer>,
+	timestamp_source: Box<TimestampSource>,
+}
+
+impl Logging {
+	/// Creates the logging object
+	pub fn new(logs_serializer: Arc<LogsSerializer>, timestamp_source: Box<TimestampSource>) -> Self {
+		let logging = Logging {
+			logs: RwLock::new(HashMap::new()),
+			logs_serializer,
+			timestamp_source,
+		};
+		if let Err(err) = logging.read_logs() {
+			warn!(target: "privatetx", "Cannot read logs: {:?}", err);
+		}
+		logging
+	}
+
+	/// Retrieves log for the corresponding tx hash
+	pub fn tx_log(&self, tx_hash: &H256) -> Option<TransactionLog> {
+		self.logs.read().get(&tx_hash).cloned()
+	}
+
+	/// Logs the creation of private transaction
+	pub fn private_tx_created<'a>(&self, tx_hash: &H256, validators: &Vec<Address>) {
+		let mut validator_logs = Vec::new();
+		for account in validators {
+			validator_logs.push(ValidatorLog {
+				account: *account,
+				validated: false,
+				validation_timestamp: None,
+			});
+		}
+		let mut logs = self.logs.write();
+		if logs.len() > MAX_JOURNAL_LEN {
+			// Remove the oldest log
+			if let Some(tx_hash) = logs.values()
+				.min_by(|x, y| x.creation_timestamp.cmp(&y.creation_timestamp))
+				.map(|oldest| oldest.tx_hash)
+			{
+				logs.remove(&tx_hash);
+			}
+		}
+		logs.insert(*tx_hash, TransactionLog {
+			tx_hash: *tx_hash,
+			status: PrivateTxStatus::Created,
+			creation_timestamp: self.timestamp_source.current_timestamp(),
+			validators: validator_logs,
+			deployment_timestamp: None,
+			public_tx_hash: None,
+		});
+	}
+
+	/// Logs the obtaining of the signature for the private transaction
+	pub fn signature_added(&self, tx_hash: &H256, validator: &Address) {
```

Comment:
I don't understand the documentation and naming of this method.

To me it seems like it tries to validate a transaction and append the timestamp to it

Reply:
Documentation re-phrased